docs/proposals/202005_query_logging_rate_limiting.md

```diff
@@ -0,0 +1,126 @@
+
+## Use Cases
+
+* Controlling and limiting the amount of requests made by the internal APIs. 
+* Query logging part would help in debugging the internal requests made.
+* This provides ample of insights about possible bottle-neck of individual components of Thanos.
+
+
+## Goals of this design
+
+* Implement Query Logging for internal APIs of Thanos, StoreAPI in particular
+* The Query Logging to implement a separate logger derived from the base logger. 
+* Implement Rate limiting and tracking of queries of the internal APIs
+
+
+## Proposal
+
+### Query Logging in Thanos
+
+#### Description of the Problem
+
+The Query Logging is a feature which helps in tracking all the active queries made in a database, and helps the users in knowing/debugging what queries made were resource-intensive, and how they can use the details to further optimize the database, or from the user’s perspective, to draw out the line from making such a request.
+
+A similar feature exists in Prometheus, here is a description of an active Query Log [[1]](https://www.robustperception.io/what-queries-were-running-when-prometheus-died) -
+
+```powershell
+level=info ts=2019-08-28T14:30:09.142Z caller=main.go:329 msg="Starting Prometheus" version="(version=2.12.0, branch=HEAD, revision=43acd0e2e93f9f70c49b2267efa0124f1e759e86)"
+level=info ts=2019-08-28T14:30:09.142Z caller=main.go:330 build_context="(go=go1.12.8, user=root@7a9dbdbe0cc7, date=20190818-13:53:16)"
+level=info ts=2019-08-28T14:30:09.142Z caller=main.go:331 host_details="(Linux 4.15.0-55-generic #60-Ubuntu SMP Tue Jul 2 18:22:20 UTC 2019 x86_64 mari (none))"
+level=info ts=2019-08-28T14:30:09.142Z caller=main.go:332 fd_limits="(soft=1000000, hard=1000000)"
+level=info ts=2019-08-28T14:30:09.142Z caller=main.go:333 vm_limits="(soft=unlimited, hard=unlimited)"
+level=info ts=2019-08-28T14:30:09.143Z caller=query_logger.go:74 component=activeQueryTracker msg="These queries didn't finish in prometheus' last run:" queries="[{\"query\":\"changes(changes(prometheus_http_request_duration_seconds_bucket[1h:1s])[1h:1s])\",\"timestamp_sec\":1567002604}]"
+level=info ts=2019-08-28T14:30:09.144Z caller=main.go:654 msg="Starting TSDB ..."level=info
+```
+
+Here, we can see that this log was taken before the prometheus instance died unnaturally, and the query logged under the component activeQueryTracker, logs the respective query that did the damage. Even though it does not pin-points the exact issue that created the problem, still it is quite helpful in considering this issue as one of the potential queries that led to the unnatural death of the Prometheus instance.
+
+#### Possible Approach
```

Comment:
Let's make sure what use case we are solving with proposed solution

Reply:
Yup, revamped the structure for removing redundant content.